File: Stripe/STPPaymentCardTextFieldCell.m
//
//  STPPaymentCardTextFieldCell.m
//  Stripe
//
//  Created by Jack Flintermann on 6/16/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import "STPPaymentCardTextFieldCell.h"
#import "STPPaymentCardTextField+Private.h"

#import "UIView+Stripe_SafeAreaBounds.h"

@interface STPPaymentCardTextFieldCell()

@property (nonatomic, weak) STPPaymentCardTextField *paymentField;

@end

@implementation STPPaymentCardTextFieldCell

- (instancetype)init {
    self = [super init];
    if (self) {
        STPPaymentCardTextField *paymentField = [[STPPaymentCardTextField alloc] initWithFrame:self.bounds];
        [self.contentView addSubview:paymentField];
        _paymentField = paymentField;

        _theme = [STPTheme defaultTheme];
        [self updateAppearance];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.paymentField.frame = self.contentView.bounds;
}

- (void)setTheme:(STPTheme *)theme {
    _theme = theme;
    [self updateAppearance];
}

- (void)updateAppearance {
    self.paymentField.backgroundColor = [UIColor clearColor];
    self.paymentField.placeholderColor = self.theme.tertiaryForegroundColor;
    self.paymentField.borderColor = [UIColor clearColor];
    self.paymentField.textColor = self.theme.primaryForegroundColor;
    self.paymentField.textErrorColor = self.theme.errorColor;
    self.paymentField.font = self.theme.font;
    self.backgroundColor = self.theme.secondaryBackgroundColor;
}

- (void)setInputAccessoryView:(UIView *)inputAccessoryView {
    _inputAccessoryView = inputAccessoryView;
    self.paymentField.inputAccessoryView = inputAccessoryView;
}

- (BOOL)isEmpty {
    return self.paymentField.cardNumber.length == 0;
}

- (BOOL)becomeFirstResponder {
    return [self.paymentField becomeFirstResponder];
}

- (NSInteger)accessibilityElementCount {
    return [[self.paymentField allFields] count];
}

- (id)accessibilityElementAtIndex:(NSInteger)index {
    return [self.paymentField allFields][index];
}

- (NSInteger)indexOfAccessibilityElement:(id)element {
    NSArray *fields = [self.paymentField allFields];
    for (NSUInteger i = 0; i < [fields count]; i++) {
        if (element == fields[i]) {
            return i;
        }
    }
    return 0;
}


@end

File: Stripe/STPPaymentCardTextFieldCell.swift
//
//  STPPaymentCardTextFieldCell.swift
//  Stripe
//
//  Created by Jack Flintermann on 6/16/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

import UIKit

class STPPaymentCardTextFieldCell: UITableViewCell {
  private(set) weak var paymentField: STPPaymentCardTextField?

  var theme: STPTheme = STPTheme.defaultTheme {
    didSet {
      updateAppearance()
    }
  }

  private var _inputAccessoryView: UIView?
  override var inputAccessoryView: UIView? {
    get {
      _inputAccessoryView
    }
    set(inputAccessoryView) {
      _inputAccessoryView = inputAccessoryView
      paymentField?.inputAccessoryView = inputAccessoryView
    }
  }

  func isEmpty() -> Bool {
    return (paymentField?.cardNumber?.count ?? 0) == 0
  }

  override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
    super.init(style: style, reuseIdentifier: reuseIdentifier)
    let paymentField = STPPaymentCardTextField(frame: bounds)
    paymentField.postalCodeEntryEnabled = false
    contentView.addSubview(paymentField)
    self.paymentField = paymentField
    theme = STPTheme.defaultTheme
    updateAppearance()
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    paymentField?.frame = contentView.bounds
  }

  @objc func updateAppearance() {
    paymentField?.backgroundColor = UIColor.clear
    paymentField?.placeholderColor = theme.tertiaryForegroundColor
    paymentField?.borderColor = UIColor.clear
    paymentField?.textColor = theme.primaryForegroundColor
    paymentField?.textErrorColor = theme.errorColor
    paymentField?.font = theme.font
    backgroundColor = theme.secondaryBackgroundColor
  }

  @objc override func becomeFirstResponder() -> Bool {
    return paymentField?.becomeFirstResponder() ?? false
  }

  override func accessibilityElementCount() -> Int {
    return paymentField?.allFields.count ?? 0
  }

  override func accessibilityElement(at index: Int) -> Any? {
    return paymentField?.allFields[index]
  }

  override func index(ofAccessibilityElement element: Any) -> Int {
    let fields = paymentField?.allFields
    for i in 0..<(fields?.count ?? 0) {
      if (element as? AnyHashable) == fields?[i] {
        return i
      }
    }
    return 0
  }

  required init?(coder aDecoder: NSCoder) {
    super.init(coder: aDecoder)
  }
}
